ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsMaterializedTableIQv2.java

```diff
@@ -0,0 +1,193 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams.materialization.ks;
+
+import static org.apache.kafka.streams.query.StateQueryRequest.inStore;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Streams;
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.streams.materialization.MaterializationException;
+import io.confluent.ksql.execution.streams.materialization.MaterializedTable;
+import io.confluent.ksql.execution.streams.materialization.Row;
+import io.confluent.ksql.util.IteratorUtil;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.kafka.streams.query.KeyQuery;
+import org.apache.kafka.streams.query.QueryResult;
+import org.apache.kafka.streams.query.RangeQuery;
+import org.apache.kafka.streams.query.StateQueryRequest;
+import org.apache.kafka.streams.query.StateQueryResult;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+
+/**
+ * Kafka Streams impl of {@link MaterializedTable}.
+ */
+class KsMaterializedTableIQv2 implements MaterializedTable {
+
+  private final KsStateStore stateStore;
+
+  KsMaterializedTableIQv2(final KsStateStore store) {
+    this.stateStore = Objects.requireNonNull(store, "store");
+  }
+
+  @Override
+  public Optional<Row> get(
+      final GenericKey key,
+      final int partition
+  ) {
+    try {
+      final KeyQuery<GenericKey, ValueAndTimestamp<GenericRow>> query = KeyQuery.withKey(key);
+      final StateQueryRequest<ValueAndTimestamp<GenericRow>>
+          request = inStore(stateStore.getStateStoreName())
+          .withQuery(query)
+          .withPartitions(ImmutableSet.of(partition));
+      final StateQueryResult<ValueAndTimestamp<GenericRow>>
+          result = stateStore.getKafkaStreams().query(request);
+
+      final Map<Integer, QueryResult<ValueAndTimestamp<GenericRow>>> queryResultMap =
+          result.getPartitionResults();
+      if (queryResultMap.size() != 1) {
+        throw new MaterializationException("The query did not return exactly one partition result: "
+                + result.getPartitionResults());
+      }
+      if (!queryResultMap.containsKey(partition)) {
+        throw new MaterializationException(
+            "The query did not return the requested partition result: "
+                + result.getPartitionResults());
+      }
+      final QueryResult<ValueAndTimestamp<GenericRow>> queryResult =
+          queryResultMap.get(partition);
+      if (queryResult.isFailure()) {
+        throw new MaterializationException(queryResult.getFailureMessage());
+      }
```

Comment:
Note: you could just not handle this case. If you try and do `getResult` on a failed result, it already throws an exception that contains the failure reason and message:

https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/query/internals/FailedQueryResult.java#L84-L88